libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,672 @@
+        case param_s::YAW_SMAX:
+            return &attitude_control.get_rate_yaw_pid().slew_limit();    
+        case param_s::YAW_FLTT:
+            return &attitude_control.get_rate_yaw_pid().filt_T_hz();
+        case param_s::YAW_FLTD:
+            return &attitude_control.get_rate_yaw_pid().filt_D_hz();
+        case param_s::YAW_FLTE:
+            return &attitude_control.get_rate_yaw_pid().filt_E_hz();
+        case param_s::YAW_FF:
+            return &attitude_control.get_rate_roll_pid().ff();
+        default:
+            GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "INTERNAL ERROR - get_param_pointer - param was %d", uint8_t(param));
+            // INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            return nullptr;
+    }
+}
+
+float AP_Quicktune::get_param_value(AP_Quicktune::param_s param){
+    AP_Float *ptr = get_param_pointer(param);
+    if (ptr != nullptr) {
+        return ptr->get();
+    }
+    GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "INTERNAL ERROR - get_param_value - param was %d", uint8_t(param));
+    return 0.0;
+}
+
+void AP_Quicktune::set_param_value(AP_Quicktune::param_s param, float value){
+    AP_Float *ptr = get_param_pointer(param);
+    if (ptr != nullptr) {
+       ptr->set(value);
+       return;
+    }
+    GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "INTERNAL ERROR - set_param_value - param was %d", uint8_t(param));
+    return;
+}
+
+void AP_Quicktune::set_and_save_param_value(AP_Quicktune::param_s param, float value){
+    AP_Float *ptr = get_param_pointer(param);
+    if (ptr != nullptr) {
+       ptr->set_and_save(value);
+       return;
+    }
+    GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "INTERNAL ERROR - set_param_value - param was %d", uint8_t(param));
+    return;
+}
+
+AP_Quicktune::axis_names AP_Quicktune::get_axis(AP_Quicktune::param_s param)
+{
+    if (param < param_s::PIT_P){
+        return axis_names::RLL;
+    } else if (param < param_s::YAW_P){
+        return axis_names::PIT;
+    } else if (param < param_s::END){
+        return axis_names::YAW;
+    } else {
+        return axis_names::END;
+    }
+}
+
+const char* AP_Quicktune::get_axis_name(AP_Quicktune::axis_names axis)
+{
+    switch (axis)
+    {
+        case axis_names::RLL:
+            return "Roll";
+        case axis_names::PIT:
+            return "Pitch";
+        case axis_names::YAW:
+            return "Yaw";
+        default:
+            GCS_SEND_TEXT(MAV_SEVERITY_EMERGENCY, "INTERNAL ERROR - get_axis_name - axis was %d", uint8_t(axis));
+            return "UNK";
+    }
+}
+
+float AP_Quicktune::gain_limit(AP_Quicktune::param_s param)
+{
+    if (get_axis(param) == axis_names::YAW){
+        if (param == param_s::YAW_P){
+            return yaw_p_max;
+        }
+        if (param == param_s::YAW_D){
+            return yaw_d_max;
+        }
+    }
+   return 0.0;
+}
+
+void AP_Quicktune::write_quik(float srate, float gain, AP_Quicktune::param_s param)
+{
+#ifdef HAL_LOGGING_ENABLED
```

Comment:
```suggestion
#ifdef HAL_LOGGING_ENABLED
void AP_Quicktune::write_quik(float srate, float gain, AP_Quicktune::param_s param)
{
```

Reply:
Don't quite a few places in the code ifdef only the contents of the function?

Reply:
indeed, this would make the code less readable for a non-existent use case (ie. quicktune enabled with no logging)

Reply:
It's the pattern that's important - we've probably saved a feature's worth of flash in places which *can* be crossed by omitting the entire function.  I'm not that fussed, 'though.

... there's still the bug that I missed which is that `#ifdef` is being used rather than `#if`